File: 4.Authentication and authorization for backend engineers/code/typescript/16_authorization_rbac_11.ts
export interface Subject {
    id: string;
    dept: string;
    role: string;
}

export interface Resource {
    ownerID: string;
    dept: string;
    archived: boolean;
}

// Policy: can `s` perform `action` on `r`, given the environment?
export function canEdit(s: Subject, r: Resource, hour: number): boolean {
    if (r.archived) {
        return false;
    }
    const sameDept = s.dept === r.dept;
    const owns = s.id === r.ownerID;
    const businessHours = hour >= 9 && hour < 18;
    // owner OR same-department editor, only in business hours
    return (owns || (sameDept && s.role === 'editor')) && businessHours;
}

// usage inside a handler, after auth has populated the subject:
// if (!canEdit(subj, doc, new Date().getHours())) {
//     res.status(403).send('forbidden');
// }

File: 4.Authentication and authorization for backend engineers/code/rust/16_authorization_rbac_11.rs
pub struct Subject {
    pub id: String,
    pub dept: String,
    pub role: String,
}

pub struct Resource {
    pub owner_id: String,
    pub dept: String,
    pub archived: bool,
}

// Policy: can `s` perform `action` on `r`, given the environment?
pub fn can_edit(s: &Subject, r: &Resource, hour: u32) -> bool {
    if r.archived {
        return false;
    }
    let same_dept = s.dept == r.dept;
    let owns = s.id == r.owner_id;
    let business_hours = hour >= 9 && hour < 18;
    
    // owner OR same-department editor, only in business hours
    (owns || (same_dept && s.role == "editor")) && business_hours
}

// usage inside a handler, after auth has populated the subject:
// use chrono::Local;
// if !can_edit(&subj, &doc, Local::now().hour()) {
//     return HttpResponse::Forbidden().body("forbidden");
// }
